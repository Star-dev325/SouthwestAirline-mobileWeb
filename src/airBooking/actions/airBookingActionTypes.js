import { actionCreator } from 'src/shared/redux/actionCreator';

const { createTypes, createApiActions } = actionCreator('airBooking');

const types = {
  sync: [
    'CALENDAR_STRIP',
    'CLEAR_ACCOUNT_INFO',
    'CLEAR_MULTI_SELECT_BOUND',
    'CLEAR_SELECTED_FREQUENT_TRAVELERS',
    'CLEAR_SPECIAL_ASSISTANCE',
    'CLEAR_SPLIT_PAY_OPTIONS_LIST',
    'DELETE_CURRENT_SEARCH_REQUEST',
    'FLIGHT_PRICING_RESUME_AFTER_LOGIN',
    'FLIGHT_PRICING_SET_HAS_UPSELL_ERROR',
    'GENERATE_PASSENGER_PAGE_INFO',
    'GENERATE_PURCHASE_SUMMARY_PAGE',
    'LOAD_FREQUENT_TRAVELER_PAGE',
    'LOW_FARE_CALENDAR_SELECT_INBOUND_DATE',
    'LOW_FARE_CALENDAR_SELECT_OUTBOUND_DATE',
    'PREFILL_PASSENGER_INFO',
    'REMOVE_SELECTED_FREQUENT_TRAVELER_PAX_ID',
    'RESET_AIR_BOOKING_FLOW_DATA',
    'RESET_CONTACT_METHOD',
    'RESET_FLIGHT_PRICING_PAGE_RESPONSE',
    'RESET_FLIGHT_SEARCH_REQUEST',
    'RESET_PASSENGER_PASSPORT',
    'RESET_PASSENGER',
    'RESET_PAYMENT_INFO',
    'RESET_SELECTED_PRODUCTS',
    'RESET_SPLIT_PAY_TERMS_AND_CONDITIONS',
    'RESUME_AIR_BOOKING_FLOW_DATA',
    'SAVE_CHASE_SESSION_ID',
    'SAVE_EARLY_BIRD_SELECTED',
    'SAVE_PASSENGER_COUNT',
    'SAVE_PASSENGER_PASSPORT',
    'SAVE_PASSENGER',
    'SAVE_PAYMENT_INFO',
    'SAVE_SELECTED_FLIGHT',
    'SAVE_SELECTED_PRODUCTS',
    'SAVE_SPLIT_PAY_TERMS_AND_CONDITIONS',
    'SAVE_TRAVEL_FUNDS_ADDRESS',
    'SELECT_FLIGHT_PRODUCT',
    'SELECTED_FREQUENT_TRAVELER',
    'SET_EARLY_BIRD_ELIGIBILITY',
    'SET_EARLY_BIRD_PRICING_DIFFERENCE',
    'SET_EARLY_BIRD_PRICING_TOKEN',
    'SET_ELIGIBLE_EXPRESS_CHECKOUT',
    'SET_EXPRESS_CHECKOUT_FROM_PASSENGER_PAGE',
    'SET_INTERNATIONAL_BOOKING_FLAG',
    'SET_IS_EXPRESS_CHECKOUT',
    'SET_IS_LOW_FARE_CALENDAR_ENABLED',
    'SET_IS_SPLIT_PAY_VISIBLE',
    'SET_IS_UPLIFT_VISIBLE',
    'SORT_FLIGHT_SHOPPING_PAGE_BY',
    'SPLIT_PAY_RESUME_AFTER_LOGIN',
    'UPDATE_CONTACT_METHOD',
    'UPDATE_CONTACT_TRAVEL_INFO_METHOD',
    'UPDATE_CORPORATE_BOOKING_SWITCH_INFO',
    'UPDATE_FLIGHT_INITIAL_SEARCH',
    'UPDATE_FLIGHT_SEARCH_REQUEST',
    'UPDATE_FREQUENT_TRAVELER_PAX_ID',
    'UPDATE_LOW_FARE_CALENDAR_INBOUND_ANALYTICS',
    'UPDATE_LOW_FARE_CALENDAR_OUTBOUND_ANALYTICS',
    'UPDATE_MULTI_SELECT_BOUND',
    'UPDATE_PASSENGER',
    'UPDATE_SELECTED_IRN',
    'UPDATE_SPECIAL_ASSISTANCE'
  ],
  async: [
    'FETCH_ACCOUNT_LOOK_UP',
    'FETCH_BOOKING_CONFIRMATION_PAGE',
    'FETCH_CONFIRMATION_PAGE_PLACEMENTS',
    'FETCH_EARLY_BIRD_IN_PATH_INFO',
    'FETCH_FLIGHT_PRICING_PAGE',
    'FETCH_FLIGHT_SELECT_FARE_PAGE_PLACEMENTS',
    'FETCH_FLIGHT_SELECT_PAGE_PLACEMENTS',
    'FETCH_INDEX_PAGE_PLACEMENTS',
    'FETCH_FLIGHT_SHOPPING_MULTI_SELECT_PAGE',
    'FETCH_FLIGHT_SHOPPING_PAGE',
    'FETCH_LOW_FARE_CALENDAR',
    'FETCH_NEXT_LOW_FARE_CALENDAR_INBOUND_PAGE',
    'FETCH_NEXT_LOW_FARE_CALENDAR_OUTBOUND_PAGE',
    'FETCH_PASSENGER_VALIDATIONS',
    'FETCH_PREV_LOW_FARE_CALENDAR_INBOUND_PAGE',
    'FETCH_PREV_LOW_FARE_CALENDAR_OUTBOUND_PAGE',
    'FETCH_PRICE_PAGE_PLACEMENTS',
    'FETCH_PURCHASE_PAGE_PLACEMENTS',
    'FETCH_RECENT_SEARCH_PAGE',
    'FETCH_SAVED_CREDIT_CARDS_AND_PASSENGER_INFO',
    'FETCH_SPLIT_PAY_OPTIONS_LIST',
    'FETCH_SPLIT_PAY_PAGE_PLACEMENTS'
  ]
};

export const apiActionCreator = createApiActions;
export default createTypes(types);
