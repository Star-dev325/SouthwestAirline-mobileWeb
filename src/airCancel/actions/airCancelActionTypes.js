import { actionCreator } from 'src/shared/redux/actionCreator';

const { createTypes, createApiActions } = actionCreator('airCancel');

const types = {
  sync: ['RESET_FLOW_DATA', 'SET_FLOW_STATUS', 'SELECT_BOUND_ANALYTICS'],
  async: [
    'FETCH_CANCEL_BOUND_CONFIRMATION',
    'FETCH_CONFIRMATION',
    'FETCH_FLIGHT_AND_CANCEL_BOUND_WITH_SEARCH_TOKEN',
    'FETCH_REFUND_QUOTE_FOR_CANCEL_BOUND',
    'FETCH_RESERVATION_FOR_CANCEL_BOUND',
    'FETCH_SPLIT_PNR_RESERVATION_FOR_CANCEL'
  ]
};

export const apiActionCreator = createApiActions;
export default createTypes(types);
