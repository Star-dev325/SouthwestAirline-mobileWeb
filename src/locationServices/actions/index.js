import _ from 'lodash';
import { getPermissionErrorMessageBasedOnPhoneType } from 'src/locationServices/helpers/messageHelper';
import BrowserObject from 'src/shared/helpers/browserObject';
import { POSITION_ERROR, SITA_FAILURE, NO_NEAREST_SWA_AIRPORT } from 'src/shared/constants/errorCodes';
import i18n from '@swa-ui/locale';
import { containsApiErrorCodes } from 'src/shared/helpers/errorCodesHelper';
import * as LocationServiceApi from 'src/shared/api/locationServiceApi';
import LocationServicesActionTypes, {
  apiActionCreator
} from 'src/locationServices/actions/locationServicesActionTypes';
import { store } from 'src/shared/redux/createStore';

const { navigator } = BrowserObject;
const { PERMISSION_DENIED } = POSITION_ERROR;
const currentPositionOptions = { timeout: 90000, maximumAge: 0 };
const { fetchLocation, fetchLocationSuccess, fetchLocationFailed } = apiActionCreator(
  LocationServicesActionTypes.LOCATION_SERVICE__FETCH_LOCATION
);

export const getPhoneLocation = () => {
  store.dispatch(fetchLocation());

  return new Promise((resolve, reject) => {
    navigator.geolocation.getCurrentPosition(
      (response) => {
        if (response) {
          resolve(response);
          store.dispatch(fetchLocationSuccess());
        }
      },
      (error) => {
        if (error) {
          if (!navigator.onLine) {
            reject(error);
          }
          const errorCode = _.get(error, 'code');
          const isPermissionDeniedError = errorCode === PERMISSION_DENIED;

          if (isPermissionDeniedError) {
            reject(new Error(getPermissionErrorMessageBasedOnPhoneType()));
          } else {
            reject(new Error(i18n('LOCATION_SERVICES__UNABLE_TO_RETRIEVE_LOCATION')));
          }
          store.dispatch(fetchLocationFailed());
        }
      },
      currentPositionOptions
    );
  });
};

export const getPhoneLocationForWebView = () => {
  store.dispatch(fetchLocation());

  return new Promise((resolve, reject) =>
    navigator.geolocation.getCurrentPosition(
      (response) => {
        if (response) {
          resolve(response);
          store.dispatch(fetchLocationSuccess());
        }
      },
      (error) => {
        reject(error);
        store.dispatch(fetchLocationFailed());
      },
      currentPositionOptions
    )
  );
};

const nearestAirportWithCoordinatesSuccess = (response) => response;
const nearestAirportWithCoordinatesError = (err) => {
  if (shouldOverrideLocationError(err)) {
    return new Error(i18n('SHARED__ERROR_MESSAGES__LOCATION_UNAVAILABLE'));
  } else {
    return err;
  }
};

export const fetchNearestAirportWithCoordinates = (longitude, latitude) => {
  store.dispatch(fetchLocation());

  return LocationServiceApi.fetchNearestAirportWithCoordinates(longitude, latitude)
    .then((response) => {
      store.dispatch(fetchLocationSuccess());

      return nearestAirportWithCoordinatesSuccess(response);
    })
    .catch((response) => {
      store.dispatch(fetchLocationFailed(nearestAirportWithCoordinatesError(response)));
    });
};
const shouldOverrideLocationError = (err) => {
  const containsSpecificErrorCode = containsApiErrorCodes(err, SITA_FAILURE, NO_NEAREST_SWA_AIRPORT);
  const isOnLine = navigator.onLine;

  return isOnLine && !containsSpecificErrorCode;
};
