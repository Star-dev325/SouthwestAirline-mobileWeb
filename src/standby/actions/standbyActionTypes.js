import { actionCreator } from 'src/shared/redux/actionCreator';

const { createTypes, createApiActions } = actionCreator('standby');

const types = {
  sync: ['SAVE_IS_REVENUE'],
  async: ['FETCH_CHECK_STANDBY_NEAR_AIRPORT']
};

export const apiActionCreator = createApiActions;
export default createTypes(types);
