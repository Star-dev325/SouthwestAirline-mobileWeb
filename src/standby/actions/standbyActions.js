// @flow
import { push } from 'connected-react-router';
import FlowStatusActions from 'src/shared/actions/flowStatusActions';
import { getCurrentAppFlow } from 'src/shared/selectors/appSelector';
import * as StandbyApi from 'src/shared/api/standbyApi';
import { STATUS } from 'src/shared/constants/flowConstants';
import SharedConstants from 'src/shared/constants/sharedConstants';
import { buildPathWithParamAndQuery } from 'src/shared/helpers/pathUtils';
import StandbyActionTypes, { apiActionCreator } from 'src/standby/actions/standbyActionTypes';

import type { ApiErrorType } from 'src/shared/flow-typed/shared.types';
import type { Dispatch as ReduxDispatch } from 'redux';

const { STANDBY__SAVE_IS_REVENUE, STANDBY__FETCH_CHECK_STANDBY_NEAR_AIRPORT } = StandbyActionTypes;

export const shouldRedirectToHomePage = (state: *, error: ApiErrorType) => {
  const pathname = getCurrentAppFlow(state);

  return error?.responseJSON?.code === 400308278 && 
    (pathname === SharedConstants.APP_FLOWS.SAME_DAY || pathname === SharedConstants.APP_FLOWS.STANDBY);
};

const { fetchCheckStandbyNearAirport, fetchCheckStandbyNearAirportSuccess, fetchCheckStandbyNearAirportFailed } =
  apiActionCreator(STANDBY__FETCH_CHECK_STANDBY_NEAR_AIRPORT, { shouldRedirectToHomePage });

export const checkStandbyNearAirport =
  (query: *, shouldPushToStandby: boolean, isRevenue: boolean) => (dispatch: ReduxDispatch<*>) => {
    dispatch(fetchCheckStandbyNearAirport(query));
    dispatch(FlowStatusActions.setFlowStatus('standby', STATUS.INITIAL));

    return StandbyApi.fetchStandbyList(query)
      .then((apiResponse) => {
        dispatch(fetchCheckStandbyNearAirportSuccess(apiResponse));
        dispatch(FlowStatusActions.setFlowStatus('standby', STATUS.IN_PROGRESS));

        if (shouldPushToStandby) {
          dispatch(saveStandbyIsRevenueForAnalytics(isRevenue));
          dispatch(push(buildPathWithParamAndQuery('/standby', null, query)));
        }
      })
      .catch((error) => dispatch(fetchCheckStandbyNearAirportFailed(error)));
  };

export const checkEnhancedStandbyNearAirport =
  (query: *, shouldPushToStandby: boolean, isRevenue: boolean, state: {firstName: string, lastName: string, recordLocator: string}) => (dispatch: ReduxDispatch<*>) => {
    dispatch(fetchCheckStandbyNearAirport(query));
    dispatch(FlowStatusActions.setFlowStatus('standby', STATUS.INITIAL));

    return StandbyApi.fetchEnhancedStandbyList(query)
      .then((apiResponse) => {
        dispatch(fetchCheckStandbyNearAirportSuccess(apiResponse));
        dispatch(FlowStatusActions.setFlowStatus('standby', STATUS.IN_PROGRESS));

        if (shouldPushToStandby) {
          dispatch(saveStandbyIsRevenueForAnalytics(isRevenue));
          dispatch(push(buildPathWithParamAndQuery('/standby', null, query), { ...state }));
        }
      })
      .catch((error) => dispatch(fetchCheckStandbyNearAirportFailed(error)));
  };

const saveStandbyIsRevenueForAnalytics = (isRevenue: boolean) => ({
  type: STANDBY__SAVE_IS_REVENUE,
  isRevenue
});
