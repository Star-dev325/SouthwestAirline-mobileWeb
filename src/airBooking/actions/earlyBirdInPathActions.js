// @flow

import * as purchaseSummaryPageHelper from 'src/airBooking/helpers/purchaseSummaryPageHelper';
import * as FlightBookingApi from 'src/shared/api/flightBookingApi';
import AirBookingActionTypes, { apiActionCreator } from 'src/airBooking/actions/airBookingActionTypes';
import { setEarlyBirdEligibility, setEarlyBirdPricingDifference } from 'src/airBooking/actions/airBookingActions';

import type { CurrencySuit, EarlyBirdPricing, PassengerInfos } from 'src/shared/flow-typed/shared.types';
import type { EarlyBirdEligibility } from 'src/airBooking/flow-typed/airBooking.types';

const { fetchEarlyBirdInPathInfo, fetchEarlyBirdInPathInfoSuccess, fetchEarlyBirdInPathInfoFailed } = apiActionCreator(
  AirBookingActionTypes.AIR_BOOKING__FETCH_EARLY_BIRD_IN_PATH_INFO
);

export const fetchEarlybirdPricing = (
  earlyBirdPricing: EarlyBirdPricing,
  passengerInfos: PassengerInfos,
  currencyType: CurrencySuit,
  earlyBirdPricingToken?: string
) => {
  const earlybirdInPathRequest = purchaseSummaryPageHelper.transformToEarlybirdInPathRequest(
    earlyBirdPricing,
    passengerInfos,
    currencyType,
    earlyBirdPricingToken
  );

  return (dispatch: *) => {
    dispatch(fetchEarlyBirdInPathInfo());

    return FlightBookingApi.retrieveEarlyBirdInPathInfo(earlybirdInPathRequest)
      .then((response) => {
        const { earlyBirdPricingDifference = null, earlyBirdEligibility = null } = response || {};

        earlyBirdEligibility && dispatch(setEarlyBirdEligibility(earlyBirdEligibility));
        earlyBirdPricingDifference && dispatch(setEarlyBirdPricingDifference(earlyBirdPricingDifference));

        return dispatch(fetchEarlyBirdInPathInfoSuccess(response));
      })
      .catch(() => dispatch(fetchEarlyBirdInPathInfoFailed()));
  };
};

export const saveEarlyBirdEligibility = (earlyBirdEligibility: EarlyBirdEligibility) => ({
  type: AirBookingActionTypes.AIR_BOOKING__FETCH_EARLY_BIRD_IN_PATH_INFO_SUCCESS,
  response: { earlyBirdEligibility }
});
