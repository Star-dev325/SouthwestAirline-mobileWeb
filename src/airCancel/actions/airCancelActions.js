// @flow
import i18n from '@swa-ui/locale';
import { goBack, push } from 'connected-react-router';
import _ from 'lodash';
import airCancelActionTypes, { apiActionCreator } from 'src/airCancel/actions/airCancelActionTypes';
import { AIR_CANCEL_SPLIT_PNR_FLOW_NAME, ERRORS } from 'src/airCancel/constants/airCancelConstants';
import { hideDialog, showDialog } from 'src/shared/actions/dialogActions';
import FlowStatusActions from 'src/shared/actions/flowStatusActions';
import { asyncActionFinish } from 'src/shared/actions/sharedActions';
import { raiseSatelliteEvent } from 'src/shared/analytics/helpers/analyticsEventHelper';
import * as AirCancelApi from 'src/shared/api/airCancelApi';
import { STATUS } from 'src/shared/constants/flowConstants';
import RefundTypes from 'src/shared/constants/refundTypes';
import BrowserObject from 'src/shared/helpers/browserObject';
import { playHapticFeedback } from 'src/shared/helpers/hapticFeedbackHelper';
import { buildPathWithParamAndQuery, transformSearchToQuery } from 'src/shared/helpers/pathUtils';
import { getRefundQuoteRequestData } from 'src/shared/helpers/selectPassengersHelper';
import { getNormalizedRoute } from 'src/shared/helpers/urlHelper';
import { generateDialogConfigFromError } from 'src/shared/transformers/dialogTransformer';
import { retrieveFlightReservation } from 'src/viewReservation/actions/viewReservationActions';

import type { Dispatch as ReduxDispatch } from 'redux';
import type {
  CancelBoundConfirmationRequestType,
  CancelBoundRefundQuoteRequestType
} from 'src/airCancel/flow-typed/airCancel.types';
import type { Dispatch as ThunkDispatch } from 'src/shared/flow-typed/shared.types';

const { BACK_TO_ORIGINAL_PAYMENT } = RefundTypes;

const {
  AIR_CANCEL__FETCH_CANCEL_BOUND_CONFIRMATION,
  AIR_CANCEL__FETCH_CONFIRMATION,
  AIR_CANCEL__FETCH_REFUND_QUOTE_FOR_CANCEL_BOUND,
  AIR_CANCEL__FETCH_RESERVATION_FOR_CANCEL_BOUND,
  AIR_CANCEL__FETCH_SPLIT_PNR_RESERVATION_FOR_CANCEL,
  AIR_CANCEL__RESET_FLOW_DATA,
  AIR_CANCEL__SELECT_BOUND_ANALYTICS
} = airCancelActionTypes;

const { location } = BrowserObject;

const { fetchReservationForCancelBound, fetchReservationForCancelBoundSuccess, fetchReservationForCancelBoundFailed } =
  apiActionCreator(AIR_CANCEL__FETCH_RESERVATION_FOR_CANCEL_BOUND);

const { fetchRefundQuoteForCancelBound, fetchRefundQuoteForCancelBoundSuccess, fetchRefundQuoteForCancelBoundFailed } =
  apiActionCreator(AIR_CANCEL__FETCH_REFUND_QUOTE_FOR_CANCEL_BOUND);

const { fetchCancelBoundConfirmation, fetchCancelBoundConfirmationSuccess, fetchCancelBoundConfirmationFailed } =
  apiActionCreator(AIR_CANCEL__FETCH_CANCEL_BOUND_CONFIRMATION);

export const selectBoundAnalytics = () => ({
  type: AIR_CANCEL__SELECT_BOUND_ANALYTICS
});

export const { fetchConfirmation, fetchConfirmationSuccess, fetchConfirmationFailed } =
  apiActionCreator(AIR_CANCEL__FETCH_CONFIRMATION);

export const {
  fetchSplitPnrReservationForCancel,
  fetchSplitPnrReservationForCancelFailed,
  fetchSplitPnrReservationForCancelSuccess
} = apiActionCreator(AIR_CANCEL__FETCH_SPLIT_PNR_RESERVATION_FOR_CANCEL);

export const resetAirCancelFlowData = () => ({
  type: AIR_CANCEL__RESET_FLOW_DATA
});

export const retrieveReservationForCancelBound =
  (cancelLink: Link, isLoggedIn: boolean) =>
    (dispatch: ReduxDispatch<*>): Promise<*> => {
      dispatch(resetAirCancelFlowData());
      dispatch(fetchReservationForCancelBound(cancelLink));

      return AirCancelApi.retrieveReservationForCancel(cancelLink, isLoggedIn)
        .then((response) => {
          dispatch(fetchReservationForCancelBoundSuccess(response));
          raiseSatelliteEvent('cancel choose flight');

          return response;
        })
        .catch((error) => {
          dispatch(fetchReservationForCancelBoundFailed(error));
          throw error;
        });
    };

// TODO: remove useHardCodedRoute once getNormalizedRoute is added to standby flow (MOB-125381)
export const retrieveRefundQuoteForCancelBound =
  (request: CancelBoundRefundQuoteRequestType, navigateToAirCancelQuotePage: boolean, isLoggedIn: boolean, useHardCodedRoute?: boolean) =>
    (dispatch: ReduxDispatch<*>): Promise<*> => {
      dispatch(fetchRefundQuoteForCancelBound(request));

      return AirCancelApi.retrieveRefundQuoteAndConfirmationForCancelBound(request, isLoggedIn)
        .then((response) => {
          dispatch(fetchRefundQuoteForCancelBoundSuccess(response));
          
          if (navigateToAirCancelQuotePage) {
            const { searchToken } = transformSearchToQuery(location?.search);
            const queryParams = searchToken ? { searchToken } : {};

            if (useHardCodedRoute) {
              dispatch(push(`/air/cancel/${response.cancelRefundQuotePage.recordLocator}`));
            } else {
              dispatch(
                push(
                  buildPathWithParamAndQuery(getNormalizedRoute({ routeName: 'airCancelRefundQuote' }), {
                    recordLocator: response.cancelRefundQuotePage.recordLocator
                  }, queryParams)
                )
              );
            }
          }

          return response;
        })
        .catch((error) => {
          dispatch(fetchRefundQuoteForCancelBoundFailed(error));
        });
    };

export const cancelReservationByBounds =
  (request: CancelBoundConfirmationRequestType, isLoggedIn: boolean) =>
    (dispatch: *): Promise<*> => {
      dispatch(fetchCancelBoundConfirmation(request));

      return AirCancelApi.retrieveRefundQuoteAndConfirmationForCancelBound(request, isLoggedIn)
        .then((response) => {
          const { searchToken } = transformSearchToQuery(location?.search);
          const queryParams = searchToken ? { searchToken } : {};

          dispatch(fetchCancelBoundConfirmationSuccess(response));
          dispatch(FlowStatusActions.setFlowStatus('airCancel', STATUS.COMPLETED));
          dispatch(
            push(
              buildPathWithParamAndQuery(getNormalizedRoute({ routeName: 'refundSummary' }), {
                recordLocator: response.cancelBoundConfirmationPage.recordLocator
              }, queryParams)
            )
          );
        })
        .catch((error) => {
          const errorCode = _.get(error, 'responseJSON.code');

          if (errorCode === ERRORS.CANCEL_BOUND_REFUND_QUOTE_MISMATCH) {
            const dialogConfig = generateDialogConfigFromError(error);

            dispatch(
              showDialog({
                name: 'air-cancel-quote-mismatch-message',
                buttons: [
                  {
                    label: i18n('SHARED__BUTTON_TEXT__OK'),
                    onClick: () => {
                      dispatch(hideDialog()).then(() => {
                        dispatch(goBack());
                      });
                    }
                  }
                ],
                ...dialogConfig
              })
            );
            error.$customized = true;
            dispatch(fetchCancelBoundConfirmationFailed(error));
          } else {
            dispatch(fetchCancelBoundConfirmationFailed(error));
          }
        });
    };

export const cancelReservation = (cancelLink: Link, recordLocator: string, formData: *, isLoggedIn: boolean) => {
  const { refundMethod, emailReceiptTo } = formData;
  const refundRequested = refundMethod === BACK_TO_ORIGINAL_PAYMENT;
  const cancelQuery = _.get(cancelLink, 'query');

  const newQueryValues = {};

  emailReceiptTo && _.set(newQueryValues, 'query["receipt-email"]', emailReceiptTo);
  cancelQuery['refund-requested'] !== null && _.set(newQueryValues, 'query["refund-requested"]', refundRequested);

  return (dispatch: ReduxDispatch<*>): Promise<*> => {
    const cancelReservationLink = _.merge({}, cancelLink, newQueryValues);

    dispatch(fetchConfirmation(cancelReservationLink));

    return AirCancelApi.cancelReservation(cancelReservationLink, isLoggedIn)
      .then((response) => {
        const { searchToken } = transformSearchToQuery(location?.search);
        const queryParams = searchToken ? { searchToken } : {};

        dispatch(fetchConfirmationSuccess(response));
        dispatch(FlowStatusActions.setFlowStatus('airCancel', STATUS.COMPLETED));
        dispatch(
          push(buildPathWithParamAndQuery(getNormalizedRoute({ routeName: 'refundSummary' }), { recordLocator }, queryParams))
        );
        playHapticFeedback();
      })
      .catch((error) => dispatch(fetchConfirmationFailed(error)));
  };
};

export const getSplitPnrReservationForCancel =
  (link: Link) =>
    (dispatch: *, getState: () => *): Promise<*> => {
      dispatch(fetchSplitPnrReservationForCancel(link));
      dispatch(FlowStatusActions.setFlowStatus(AIR_CANCEL_SPLIT_PNR_FLOW_NAME, STATUS.IN_PROGRESS));

      return AirCancelApi.retrieveSplitPnrReservation(link)
        .then((response) => {
          const state = _.cloneDeep(getState());
          const isUserLoggedIn = _.get(state, 'app.account.isLoggedIn', false);
          const recordLocator = _.get(response, 'viewForCancelBoundPage.recordLocator', '');
          const refundQuoteLinkObject = _.get(response, 'viewForCancelBoundPage._links.refundQuote', {});
          const showBoundSelection = _.get(response, 'viewForCancelBoundPage._meta.showBoundSelection');

          dispatch(FlowStatusActions.setFlowStatus(AIR_CANCEL_SPLIT_PNR_FLOW_NAME, STATUS.COMPLETED));

          if (showBoundSelection) {
            const { searchToken } = transformSearchToQuery(location?.search);

            const queryParams = searchToken ? { searchToken } : {};

            dispatch(fetchSplitPnrReservationForCancelSuccess(response));
            dispatch(
              push(buildPathWithParamAndQuery(getNormalizedRoute({ routeName: 'selectBound' }), { recordLocator }, queryParams))
            );
          } else {
            const refundQuoteRequestData = refundQuoteLinkObject && getRefundQuoteRequestData(refundQuoteLinkObject);

            dispatch(asyncActionFinish());
            dispatch(retrieveRefundQuoteForCancelBound(refundQuoteRequestData, true, isUserLoggedIn));
          }
        })
        .catch((error) => {
          dispatch(fetchSplitPnrReservationForCancelFailed(error));
        });
    };

export const retrieveFlightAndCancelBoundWithSearchToken =
(searchToken: string, isLoggedIn: boolean, isRefundQuote: boolean) =>
  (dispatch: ThunkDispatch): Promise<*> =>
    dispatch(retrieveFlightReservation({ passengerSearchToken: searchToken }, true))
      .then(reservationData => {
        const cancelBoundRequest = reservationData?.viewReservationViewPage?._links?.cancelBound ?? '';

        return dispatch(retrieveReservationForCancelBound(cancelBoundRequest, isLoggedIn))
          .then(cancelBoundResponse => {
            if (isRefundQuote) {
              const refundQuoteLink = cancelBoundResponse?.viewForCancelBoundPage?._links?.refundQuote ?? '';
              const quoteRequestData: CancelBoundRefundQuoteRequestType = _.merge({}, refundQuoteLink, {
                body: {
                  refundRequested: null
                }
              });

              return dispatch(retrieveRefundQuoteForCancelBound(quoteRequestData, false, isLoggedIn));
            } else {
              return cancelBoundResponse;
            }
          });
      });
