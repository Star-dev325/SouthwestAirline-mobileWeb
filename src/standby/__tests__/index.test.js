jest.mock('src/standby/actions/standbyActions');

import { render } from '@testing-library/react';
import React from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import * as StandbyActions from 'src/standby/actions/standbyActions';
import StandbyRoute from 'src/standby/index';
import configureMockStore from 'test/unit/helpers/configureMockStore';

describe('StandbyRoute', () => {
  beforeEach(() => {
    StandbyActions.checkEnhancedStandbyNearAirport.mockImplementation(() => ({ type: 'test' }));
  });

  it('should render enhanced standby Route page', () => {
    const { container } = createComponent({});

    expect(container).toMatchSnapshot();
  });

  it('should render old standby Route page', () => {
    const { container } = createComponent(
      {},
      {
        app: {
          toggles: { ENHANCED_STANDBY_LIST: false }
        }
      }
    );

    expect(container).toMatchSnapshot();
  });

  const createComponent = (props = {}, state = {}) => {
    const defaultProps = {
      match: { params: '' },
      location: {
        search: 'search',
        href: '#'
      }
    };
    const defaultState = {
      app: {
        toggles: { ENHANCED_STANDBY_LIST: true }
      },
      router: {
        location: {
          search: 'search',
          href: '#'
        }
      }
    };
    const store = configureMockStore()({ ...defaultState, ...state });

    const mergedProps = { ...defaultProps, ...props };

    return render(
      <Provider store={store}>
        <BrowserRouter>
          <StandbyRoute {...mergedProps} />
        </BrowserRouter>
      </Provider>
    );
  };
});
