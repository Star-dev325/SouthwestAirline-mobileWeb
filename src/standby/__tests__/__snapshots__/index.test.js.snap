// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StandbyRoute should render enhanced standby Route page 1`] = `
<div>
  <div
    class="stand-by"
  />
</div>
`;

exports[`StandbyRoute should render old standby Route page 1`] = `
<div>
  <div
    class="stand-by"
  />
</div>
`;
